string msg_hash
---
string new_hash
